refactor: NoteDataProvider - clarify function names, add test coverage and documentation

wildjos opened the pull request:
### Overview:
This PR improves the **NoteDataProvider** to make it more maintainable, and easier to understand. It focuses on clearer function naming and documentation, with additional test coverage for edge-cases.

---

**_Documentation_**
- Added explanatory comments to make test data creation, filtering logic, and expected results easier to understand in **note_data_provider.test**
- Added function documentation in **note_data_provider**

**_Added tests_**
- new **edge-case** tests for nullifying non-existent notes and mixed real and fake nullifiers

**_Function renaming for clarity_**
- removeNotesAfter / removeNullifiedNotes / unnullifyNotes were confusing (too similar and not descriptive).
    - unnullifyNotesAfter -> **rewindNullifiersAfterBlock**
    - removeNullifiedNotes -> **applyNullifications**
    - removeNotesAfter -> **deleteNotesAfterBlock**

- pxe_oracle_interface "removeNullifiedNotes" confusing name, renamed: 
    - removeNullifiedNotes -> **syncContractNotes**

**_Identical Reimplementation in PrivateExecutionOracle_**
- Removed **utilityFetchTaggedLogs**

**_Removed getSize() and DataProvider Interface_**
- **getSize** function no longer required in each *DataProvider class -> removed
- **DataProvider** Interface now empty -> removed 

**_Wrapping private functions_**
- Added **syncNotesAndNullifers** to wrap **rewindNullifiersAfterBlock** and **deleteNotesAfterBlock**

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -886,13 +886,25 @@ export class PXEOracleInterface implements ExecutionDataProvider {
 
-  public async removeNullifiedNotes(contractAddress: AztecAddress) {
+  /**
+   * Nullifies notes associated with a given contract that have been included onchain.
```
```suggestion
   * Nullifies notes associated with a given contract.
```
Found the end of the sentence to be confusing and think it's better without it.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -326,6 +451,24 @@ export class NoteDataProvider implements DataProvider {
+   *
+   * @returns Promise resolving to the total size in bytes of all active notes
+   */
   async getSize() {
```
I think it's a bug that it counts only active notes. Generally these functions are used to estimate db size. Would add here a TODO to fix it but would not bother fixing it as I don't know if it's even relevant now.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -326,6 +451,24 @@ export class NoteDataProvider implements DataProvider {
+   * - Returns the aggregate size in bytes
+   *
+   * Note: This only counts active notes (not nullified notes) and may be expensive
+   * for large datasets as it retrieves all notes to calculate their sizes.
```
```suggestion
```

I think this is too much AI and would inevitably just get stale as people would forget to update it. Would just drop this whole section.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -9,6 +9,25 @@ import { NoteStatus, type NotesFilter } from '@aztec/stdlib/note';
+ * - Primary stores: #notes (active), #nullifiedNotes (spent)
+ * - Lookup maps: nullifier → noteId for efficient nullification
+ * - Scope-specific indexes: per-scope multi-maps for query optimization
+ * - Block tracking: nullifiers by block number for reorg handling
```
```suggestion
 * NoteDataProvider manages the storage and retrieval of active and nullified notes.
```

The rest is basically just description of implementation which is not really valuable and is at a risk of becoming stale.

Also would not mention PXE in the first sentence as PXE is unknown from the PoV of this class - if the behavior in PXE for whatever reason changed then we would most definitely not look here to update it.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -886,13 +886,25 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+   *
+   * After identifying the relevant nullifiers, the function moves the corresponding notes
+   * from the active note store to the nullified note store (`#nullifiedNotes`) and logs each
+   * nullified note for auditing purposes.
```
We are in PXE here so talking here about stores seems out of place (it's an implementation detail from the PoV of this function).

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -64,6 +92,15 @@ export class NoteDataProvider implements DataProvider {
+   * @param scope - The AztecAddress representing the scope/user to add
+   * @returns Promise resolving to true if scope was added, false if it already existed
+   */
   public async addScope(scope: AztecAddress): Promise<boolean> {
```
@nventuro I vaguely remember scopes were supposed to be removed. Is it true?

If yes would say it makes sense to do it in this PR.

nventuro replied:
I don't recall discussing removing scopes, no. They're just not being used, or not a lot at least. Wallets will want control over these. @Thunkar have we discussed anything re. removing scopes?

wildjos replied:
This PR focusses on the function renaming and docs additions.  If scopes need removing or changing, I propose handling that in a separate PR, along with removing `recipient` and `tx-hash` from the store indexes.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +148,26 @@ export class NoteDataProvider implements DataProvider {
+   * @param blockNumber - Notes created after this block number will be deleted
+   * @returns Promise that resolves when all notes have been removed
+   */
+  public deleteNotesAfterBlock(blockNumber: number): Promise<void> {
```
I know it's unrelated to your work but I think it would make sense to make `deleteNotesAfterBlock` and `restoreNotesAfterBlock` private because calling them individually is incorrect and hence it's a complexity leak that makes the code more bug-prone.

wildjos replied:
Totally agree
My plan (see response below) is to wrap them in a single function, syncNotesAndNullifiers(), and make these two private.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -270,7 +373,29 @@ export class NoteDataProvider implements DataProvider {
+   * - Removes nullifier-to-noteId mappings from active lookup
+   *
+   * The operation is atomic - if any nullifier is not found, the entire operation fails
+   * and no notes are modified.
```
```suggestion
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -176,6 +259,26 @@ export class NoteDataProvider implements DataProvider {
+   * - Uses all available scopes if none specified in filter
+   * - Queries scope-specific indexes for efficient retrieval
+   * - For ACTIVE_OR_NULLIFIED status, also searches nullified note indexes
+   * - Applies additional filtering on retrieved candidates
+   * - Returns deduplicated results matching all filter criteria
+   *
```
```suggestion
```

benesjan reviewed (commented):
Would tone down the verbosity of AI.

In-detail description of implementations are not a useful kind of docs and it's likely to become stale.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -886,13 +886,25 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+   *
+   * @param contractAddress - The contract whose notes should be checked and nullified.
+   */
+  public async nullifyContractNotes(contractAddress: AztecAddress) {
```
Usage of 'nullify' as a verb is incorrect here. Notes are 'nullified' by emitting their nullifier - the existence of the nullifier is the nullification of the note.

What this function does would be better described as 'check notes for nullification', 'check and remove nullified notes', a general 'sync contract notes', etc.

wildjos replied:
Following conversation with @nventuro this is renamed to: **syncNoteNullifiers**

nventuro commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -270,7 +373,29 @@ export class NoteDataProvider implements DataProvider {
+   * @returns Promise resolving to array of nullified NoteDao objects
+   * @throws Error if any nullifier is not found in the active notes
+   */
+  nullifyNotes(nullifiers: InBlock<Fr>[]): Promise<NoteDao[]> {
```
The verb is again a bit problematic here as the notes have already been nullified. This could be e.g. 'process note nullifiers', 'remove notes with nullifiers', etc.

wildjos replied:
Thanks, framing it this way makes sense. I see now that using “nullify” as a verb is misleading, since the function isn’t causing notes to be nullified.

I’m thinking **applyNullifications()** would be a clearer name, reflecting that it applies nullifications to the local store without actually nullifying anything itself.

wildjos replied:
Following @benesjan comment - renamed **applyNullifiers()**

nventuro commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -126,7 +189,27 @@ export class NoteDataProvider implements DataProvider {
+   * @param synchedBlockNumber - Optional upper bound for the block range to process
+   * @returns Promise that resolves when all notes have been restored
+   */
+  public async restoreNotesAfterBlock(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
```
I had to do a double-take as I was confused by this name, particularly due to seeing `deleteNotesAfter` next to `restoreNotesAfter`. I'd consider something like `removeNullifiers` possibly?

Ultimately the store does these things, with the goal of having 'a list of notes':
 - process new notes, adding them to the list
 - process nullifiers that correspond to notes it knows of (it cannot process unknown nullifs), removing notes from the list and putting them in the nullified list
 - un-process nullifiers, restoring notes to the list
 - un-process notes, deleting them outright

'process' is an extremely generic verb so I'm not necessarily saying we should use it, but perhaps it'd make sense to use consistent naming for these actions.

wildjos replied:
I see where you’re coming from - the current names highlight different aspects of the same operation: `restoreNotesAfterBlock` signals that notes are being restored, but it doesn’t make it clear that nullifiers are also being removed.

I’d propose renaming `restoreNotesAfterBlock` to `rewindNullifiersAfterBlock`, which makes it explicit that nullifiers are being rolled back.

Since `rewindNullifiersAfterBlock` and `deleteNotesAfterBlock` are always called together in the synchronizer, we could wrap them in a new function `rollbackNotesAndNullifiers()`. 

This also addresses @benesjan's earlier point by making these functions private.

The store operations now look like: 

> - process new notes, adding them to the list - **addNotes**
> - process nullifiers that correspond to notes it knows of (it cannot process unknown nullifs), removing notes from the list and putting them in the nullified list - **applyNullifiers**
> - un-process nullifiers, restoring notes to the list - **rewindNullifiersAfterBlock**
> - un-process notes, deleting them outright - **deleteActiveNotesAfterBlock**

and the wrapper would be:
```ts
private async rollbackNotesAndNullifiers(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
    await this.rewindNullifiersAfterBlock(blockNumber, synchedBlockNumber);
    await this.deleteActiveNotesAfterBlock(blockNumber);
}
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```suggestion
      for (let i = currentBlockNumber; i <= synchedBlockNumber; i++) {
```
Given that synchedBlockNumber is always defined we can drop the maxBlockNumber here.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   *
+   * @param blockNumber - Notes created after this block number will be deleted
+   */
+  private deleteNotesAfterBlock(blockNumber: number): Promise<void> {
```
```suggestion
  #deleteActiveNotesAfterBlock(blockNumber: number): Promise<void> {
```
Just a detail but we generally use the "#" prefix for private functions (TS then automatically marks the function as private) and I like it because then in the function callsite it's clear it's a private function.

Also think it would be nice to specifically mention that it deletes only the active notes. That will then make it clearer why `rewindNullifiersAfterBlock` needs to be called first.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   * @param blockNumber - The block number to synchronize to (the reorg point)
+   * @param synchedBlockNumber - Optional upper bound for the block range to process (current tip)
+   */
+  public async syncNotesAndNullifiers(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
```
```suggestion
  public async rollbackNotesAndNullifiers(blockNumber: number, synchedBlockNumber: number): Promise<void> {
```

I think the `syncNotesAndNullifiers` name insufficiently communicates that we are dealing with reorgs - here it would make me think we are syncing new notes and nullifiers.

I also checked and the `synchedBlockNumber` is always defined so please update that everywhere (It's a tech debt from time when we didn't throw in `syncDataProvider.getBlockNumber()` function).

benesjan replied:
API-wise I find it quite strange that we have the `synchedBlockNumber` on the input. It feels like a complexity leak out of this class. It also makes me worried that it might make this class more brittle - what if there is a bug somewhere and for whatever reason we store here a note or nullifier beyond the synched block number?.

Maybe we could think of how to redesign the stores here to not need this on the input. Or maybe to have it on the input but use it only to be able to check that there are no notes and nullifers beyond that block number.

This might seem too paranoid but if there was a bug here it would be a very unpleasant thing to debug.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   * specified block number. It restores any notes that were nullified after the given block
+   * and deletes any active notes created after that block.
+   *
+   * @param blockNumber - The block number to synchronize to (the reorg point)
```
```suggestion
   * @param blockNumber - The new chain tip after reorg
```

I am being admittedly pedantic here but I didn't like the "synchronize to" because that makes me think we are going "up the chain" and not "down the chain".

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   * and deletes any active notes created after that block.
+   *
+   * @param blockNumber - The block number to synchronize to (the reorg point)
+   * @param synchedBlockNumber - Optional upper bound for the block range to process (current tip)
```
```suggestion
   * @param synchedBlockNumber - The block number up to which PXE managed to sync before the reorg happened.
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
 
-  public removeNotesAfter(blockNumber: number): Promise<void> {
+  /**
+   * Synchronizes notes and nullifiers to a specific block number.
+   *
+   * This method ensures that the state of notes and nullifiers is consistent with the
+   * specified block number. It restores any notes that were nullified after the given block
+   * and deletes any active notes created after that block.
```
```suggestion
   * Deletes notes and nullifiers that were inserted in <blockNumber + 1; synchedBlockNumber> block range. Called during chain reorgs.
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   */
+  public async syncNotesAndNullifiers(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
+    await this.rewindNullifiersAfterBlock(blockNumber, synchedBlockNumber);
+    await this.deleteNotesAfterBlock(blockNumber);
```
Above this would put a comment why the order of the function calls here is important.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+  }
+
+  /**
+   * Deletes (removes) all active notes created after the specified block number.
+   *
+   * Permanently delete notes from the active notes store, e.g. during a reorg.
+   * Note: This only affects #notes (active notes), not #nullifiedNotes.
```
```suggestion
   * Deletes all active notes created after the specified block number.
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -80,17 +103,23 @@ export class NoteDataProvider implements DataProvider {
+   * @param notes - Array of NoteDao objects to store
+   * @param scope - The scope (user/account) under which to store the notes (defaults to ZERO)
+   */
   addNotes(notes: NoteDao[], scope: AztecAddress = AztecAddress.ZERO): Promise<void> {
```
```suggestion
  addNotes(notes: NoteDao[], scope: AztecAddress): Promise<void> {
```
I checked, because the zero address there seemed weird, and we always call this with the scope defined so the zero address can be dropped.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -80,17 +103,23 @@ export class NoteDataProvider implements DataProvider {
+   * uniqueness and maintains creation order. Each note is indexed by multiple criteria
+   * for efficient retrieval.
+   *
+   * @param notes - Array of NoteDao objects to store
```
```suggestion
   * @param notes - Notes to store
```

Again being pedantic but describing the type seems unnecessary (I see it's an array of note daos)

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -886,13 +886,21 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+   *
+   * @param contractAddress - The contract whose notes should be checked and nullified.
+   */
+  public async syncContractNotes(contractAddress: AztecAddress) {
```
```suggestion
  public async syncActiveNoteNullifiers(contractAddress: AztecAddress) {
```

I don't like the original name because it would make me think that we are syncing new notes but we don't do that here - we load nullifiers for known active notes.

Not super sure my proposed name is good but I would say it might be because:
1. it having contract address on the input communicates that we are most likely doing that for a single contract,
2. it explicitly mentioning active notes communicates the function tries to load nullifiers only of known active notes.

But IDK

wildjos replied:
Changed to syncNoteNullifiers

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -270,7 +343,18 @@ export class NoteDataProvider implements DataProvider {
+   * @returns Promise resolving to array of nullified NoteDao objects
+   * @throws Error if any nullifier is not found in the active notes
+   */
+  applyNullifications(nullifiers: InBlock<Fr>[]): Promise<NoteDao[]> {
```
We have `applyNullifications` name but the function that does the reverse is called `rewindNullifiersAfterBlock`. I think I would stick with `applyNullifiers` to keep the naming style consistent.

wildjos replied:
Done

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -70,184 +135,335 @@ describe('NoteDataProvider', () => {
+  // --- Nullification / Status Tests ---
+
+  describe('Note nullification and status handling', () => {
+    /**
+     * Default note status behavior test.
+     *
+     * Tests that nullified (spent) notes are excluded by default from query results.
+     * This test:
+     * 1. Adds all notes
+     * 2. Nullifies notes belonging to the first recipient (half the notes)
+     * 3. Queries notes with no status specified (default behavior)
+     * 4. Queries notes with explicit ACTIVE status
+     * 5. Verifies both queries return the same results (only active notes)
+     */
```
This is just describing what the implementation does so would remove it from here and from the rest of the test file.

These are just unit tests so the "test title" should be able to describe well enough what we are testing.

wildjos replied:
After discussion with @nventuro on how to proceed, I will tackle this in a follow-up PR that aims will restructure note_data_provider_tests to make the tests self-describing and easier to review.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
+   * @param blockNumber - The block number to synchronize to (the reorg point)
+   * @param synchedBlockNumber - Optional upper bound for the block range to process (current tip)
+   */
+  public async syncNotesAndNullifiers(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
```
Somewhat unrelated but API-wise I find it quite strange that we have the `synchedBlockNumber` on the input. It feels like a complexity leak out of this class. It also makes me worried that it might make this class more brittle - what if there is a bug somewhere and for whatever reason we store here a note or nullifier beyond the synched block number?.

Maybe we could think of how to redesign the stores here to not need this on the input. Or maybe to have it on the input but use it only to be able to check that there are no notes and nullifers beyond that block number.

This might seem too paranoid but if there was a bug here it would be a very unpleasant thing to debug.

wildjos replied:
Following discussion with @nventuro - this will move to it's own PR to address this point.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/execution_data_provider.ts:
```diff
@@ -259,9 +259,9 @@ export interface ExecutionDataProvider {
 
   /**
-   * Removes all of a contract's notes that have been nullified from the note database.
+   * Synchronizes a contract's notes by checking for and removing nullified notes.
```
```suggestion
   * Looks for nullifiers of active contract notes and marks them as nullified in the db if a nullifier is found.
```
This doesn't synchronize notes (it doesn't load new notes), it just loads new nullifiers.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -874,13 +874,21 @@ export class PXEOracleInterface implements ExecutionDataProvider {
 
-  public async removeNullifiedNotes(contractAddress: AztecAddress) {
+  /**
+   * Synchronizes notes and nullifiers associated with a given contract.
```
```suggestion
   * Looks for nullifiers of active contract notes and marks them as nullified if a nullifier is found.
```

benesjan reviewed (approved):
Feel free to merge once you apply my suggestions 👍

wildjos committed:
refactor: NoteDataProvider - clarify function names, add test coverage and documentation

### Overview:
This PR improves the **NoteDataProvider** to make it more maintainable, and easier to understand. It focuses on clearer function naming and documentation, with additional test coverage for edge-cases.

---

**_Documentation_**
- Added explanatory comments to make test data creation, filtering logic, and expected results easier to understand in **note_data_provider.test**
- Added function documentation in **note_data_provider**

**_Added tests_**
- new **edge-case** tests for nullifying non-existent notes and mixed real and fake nullifiers

**_Function renaming for clarity_**
- removeNotesAfter / removeNullifiedNotes / unnullifyNotes were confusing (too similar and not descriptive).
    - unnullifyNotesAfter -> **rewindNullifiersAfterBlock**
    - removeNullifiedNotes -> **applyNullifications**
    - removeNotesAfter -> **deleteNotesAfterBlock**

- pxe_oracle_interface "removeNullifiedNotes" confusing name, renamed:
    - removeNullifiedNotes -> **syncContractNotes**

**_Identical Reimplementation in PrivateExecutionOracle_**
- Removed **utilityFetchTaggedLogs**

**_Removed getSize() and DataProvider Interface_**
- **getSize** function no longer required in each *DataProvider class -> removed
- **DataProvider** Interface now empty -> removed

**_Wrapping private functions_**
- Added **syncNotesAndNullifers** to wrap **rewindNullifiersAfterBlock** and **deleteNotesAfterBlock**
